### ISSUE TITLE
[QUESTION] Is it correct to roll the input_ids tensor as input to the MTP module?

### ISSUE THREAD

**Issue by wavy-jung** [NVIDIA/Megatron-LM#1609]
https://github.com/NVIDIA/Megatron-LM/blob/a845aa7e12b3a117e24c2352b9e3e60bad2e3a17/megatron/core/transformer/multi_token_prediction.py#L663-L680
The input_ids of each mtp modules is determined by roll_tensor() output of previous input_ids.
https://github.com/NVIDIA/Megatron-LM/blob/a845aa7e12b3a117e24c2352b9e3e60bad2e3a17/megatron/core/transformer/multi_token_prediction.py#L118-L124
However, I just recognized that simple roll_tensor() of a tensor is just simply predicting the previous tokens rather than the future tokens which might be different from the intended implementation from the deepseek-v3 paper.
```python
>>> a = torch.arange(11)[1:]
>>> a
tensor([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])
>>> b = torch.roll(a, -1, -1)
>>> b
tensor([ 2,  3,  4,  5,  6,  7,  8,  9, 10,  1])
>>> b.select(-1, -1).fill_(0)
tensor(0)
>>> b
tensor([ 2,  3,  4,  5,  6,  7,  8,  9, 10,  0])
# is to predict the previous 9 tokens
# while the intended behavior is to predict the 9 previous tokens + 1 future token
# e.g. [ 2,  3,  4,  5,  6,  7,  8,  9, 10,  11]
```
<img width="968" alt="Image" src="https://github.com/user-attachments/assets/2217cae8-a242-4c34-b8d1-4878794fe790" />
If I’ve misunderstood something, please let me know.

**Comment by cuichenx** [comment]
Hi @wavy-jung 
Maybe you're thinking of MTP as speculative decoding?

Currently in Megatron, MTP is only used during training to provide an extra loss signal. 
In your example, during training, there are only tokens 1-10 available.  `[ 2,  3,  4,  5,  6,  7,  8,  9, 10,  11]`, token 11 does not exist in the training data. We have to use a placeholder for the last token after rolling, and the loss mask is also set to 0.

**Comment by sbhavani** [comment]
@wavy-jung do you have still have questions about MTP?
